File: 203__Remove_Linked_List_Elements.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        dummy = ListNode(0)
        curr = dummy
        while head:
            if head.val != val:
                curr.next = head
                curr = curr.next
            head = head.next
        curr.next = None
        return dummy.next

File: Java/203__Remove_Linked_List_Elements.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode cur = head, prev = new ListNode(-1), result = null;
        prev.next = head;
        if (head == null || (head.val == val && head.next == null)) {
            return null;
        }
        while (cur != null) {
            if (cur.val == val) {
                prev.next = cur.next;
                cur = cur.next;
            } else {
                if (result == null) {
                    result = cur;
                }
                prev = cur;
                cur = cur.next;
            }
        }
        return result;
    }
}
